flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/functions/sink/SocketClientSink.java

```diff
@@ -73,13 +85,49 @@ public void intializeConnection() {
 	 *			The incoming data
 	 */
 	@Override
-	public void invoke(IN value) {
+	public void invoke(IN value) throws Exception {
 		byte[] msg = schema.serialize(value);
 		try {
 			dataOutputStream.write(msg);
 		} catch (IOException e) {
-			throw new RuntimeException("Cannot send message " + value.toString() +
-					" to socket server at " + hostName + ":" + port, e);
+			Log.error("Cannot send message " + value.toString() +
+					" to socket server at " + hostName + ":" + port + ". Caused by " + e.toString() +
+					". Trying to reconnect.");
+			retries = 0;
+			boolean success = false;
+			while ((retries < maxRetry || retryForever) && !success && isRunning){
+				try {
+
+					if (dataOutputStream != null) {
+						dataOutputStream.close();
+					}
+
+					if (client != null && !client.isClosed()) {
+						client.close();
+					}
+
+					if (!retryForever){
+						retries++;
+					}
+
+					client = new Socket(hostName, port);
+					dataOutputStream = new DataOutputStream(client.getOutputStream());
+					dataOutputStream.write(msg);
+					success = true;
+
+				} catch(IOException ee) {
+					Log.error("Reconnect to socket server and send message failed. Caused by " +
+								ee.toString() + ". Retry time(s):" + retries);
+					synchronized (this) {
+						this.wait(CONNECTION_RETRY_SLEEP);
```

Comment:
This can throw an `InterruptedException`. Should be treated by checking the termination criterion for example.